### ISSUE TITLE
Fractal _preview.hbs path adds /components/ ... !?

### ISSUE THREAD

**Issue by sabsaxo** [frctl/fractal#359]
I've specified:

`. . .
fractal.components.set( 'path', path.join( __dirname, '/publish/node_modules' ) );
. . .`

NOTE: No, it's not where I put third party node_module ...

I then want to point my '_preview.hbs' at the assets a couple of levels up (as for reasons out of my hands) ... relatively, doing this in my _preview.hbs file:

`<link rel="stylesheet" href="{{ path '../../static_dev/assets/publish/css/publish.min.css' }}">
`

I've tried all variations between:
`../../../../../../../../../../../../../static_dev/assets/publish/css/publish.min.css`
and
`/static_dev/assets/publish/css/publish.min.css`

The former just gives me:
`‼︎ 404: No matching route found for /static_dev/assets/publish/js/publish.min.js`

Whereas the latter gives me:
`‼︎ 404: No matching route found for /components/preview/static_dev/assets/publish/js/publish.min.js`

And when doing (single dot):
`./static_dev/assets/publish/js/publish.min.js`

It gives:
`‼︎ 404: No matching route found for /components/preview/static_dev/assets/publish/js/publish.min.js`
Adding /components/preview ... !?

Now, I've not set components ANYWHERE (I guess it's Fractals default value somehow), but specifically set it as above ... so WHERE does the /components/ part of the path come from?

**Comment by blaasvaer** [comment]
Anyone alive in here?

**Comment by allmarkedup** [comment]
Hey @blaasvaer, yes alive but unfortunately very busy as well... sadly Fractal does not pay any bills so family and client work tends to have to take priority. I wish it wasn't like that but until people want to pay for open source software this is the situation, unfortunately.

I see you've opened a number of issues (this one, plus #363 and #360) around this - I assume that they are related. As I'm not sure I 100% understand your setup let me first try and better explain how Fractal expects you to have things configured, just to make sure there is no misunderstanding there.

First, when you specify a component directory path like so:

```
fractal.components.set( 'path', path.join( __dirname, '/publish/node_modules' ) );
```

That is the directory that the Fractal filesystem parser will look in to try and identify the components and their related files. The URLs generated in the web UI are not directly related to this directory structure at all, but instead are generated by the web UI theme and by default take the format of `/components/detail/{component-name}`.

In order to serve your [static assets](http://fractal.build/guide/web#static-assets), you can point the web UI to a 'public' directory that contains your assets. You can configure that as follows:

```
fractal.web.set('static.path', __dirname + '/path/to/public/dir');
```

Your static assets will then be served in the web UI at a path relative to the public directory you specified - i.e. a css file at `path/to/public/dir/css/main.css` will be served at the URL `/css/main.css`. You can optionally configure a URL prefix using the `fractal.web.set('static.mount', 'foobar');` config option - in this case the css file would then be served at `/foobar/css/main.css`.

In order to ensure assets paths in preview templates are correct in both the development server and the static build it's recommended that you use the [path helper](http://fractal.build/guide/core-concepts/views#handlebars-helper) although you can also use relative paths in your templates instead if this isn't an option. Let me know if this doesn't work for you and I'll try and help figure out the best way to do it.

Apologies if this is all stuff you have already tried, but I just want to check that you have everything configured properly first before digging in deeper to the specifics of your issue. Let me know if it's of any use.

**Comment by blaasvaer** [comment]
Hey Mark, good to know that you're alive – and busy.

Thanks for the elaborate answer.

As I understand the documentation, this is the stuff I have to deal with in order to get preview working (in both 'build' and, erhm 'start'):

In fractal.js I have these three lines (among others of course, and I've given them numbers to be able to reference them):

```
...
1. fractal.components.set('default.preview', '@preview');
2. fractal.web.set('static.mount', '/assets/');
3. fractal.web.set('static.path', __dirname + '/assets/');
...
```

And in my default \_preview.hbs (placed in the root of my components directory ( in my case 'node_modules' ) ):

```
...
1. <link rel="stylesheet" href="{{ path 'assets/css/fractal.min.css' }}">
2. <script src="{{ path 'assets/libs/jquery-3.1.1.min.js' }}"></script>
...
```

Now, I'm not totally clear of why I have to state the same `/assets/` in both places in the fractal.js file (as I don't really get what static.mount relates to, other than some sort of prefix to a path to your statics). So in order to understand it I fiddled around with it to see what was going on when building and previewing.

1. If I omit it completely in both places, like this:
```
2. fractal.web.set('static.mount', '');
3. fractal.web.set('static.path', __dirname + '');
```

And run `fractal build`

It pust EVERYTHING from my src directory into the build directory (basically copying it) – this means ALL third party node_modules, README.md files, package-lock.json (you get the picture) etc. creating a HUGE build with all sorts of files not needed to just run the build.

If I run `fractal start` and go to localhost:3000 and click a component, I get:

```
‼︎ 404: No matching route found for /browser-sync/browser-sync-client.js
‼︎ 404: No matching route found for /components/preview/assets/css/fractal.min.css
‼︎ 404: No matching route found for /components/preview/assets/login/css/login.min.css
‼︎ 404: No matching route found for /components/preview/assets/libs/jquery-3.1.1.min.js
‼︎ 404: No matching route found for /components/preview/assets/libs/handlebars.runtime.min.js
‼︎ 404: No matching route found for /components/preview/assets/login/js/login.min.js
‼︎ 404: No matching route found for /browser-sync/browser-sync-client.js
‼︎ 404: No matching route found for /components/preview/assets/js/fractal.templates.min.js
‼︎ 404: No matching route found for /components/preview/assets/js/fractal.min.js
‼︎ 404: No matching route found for /components/preview/assets/login/js/login.templates.min.js
‼︎ 404: No matching route found for /components/preview/assets/login/js/login.min.js
etc.
```

2. If I only add `/assets/` to the 2. line and leave the 3. line '' in fractal.js, like this:
```
2. fractal.web.set('static.mount', '/assets/');
3. fractal.web.set('static.path', __dirname + '');
```

And run `fractal build`, it puts everything from the src directory (all the third party node_modules, README.md, pakage.json files etc.) into an asset directory in the build directory (basically copying it as before).

If I run `fractal start` and go to localhost:3000 and click a component, I get the same output as before `No matching route found ...`.

3. If I only add `/assets/` to the 3. line and leave the 2. line '' in fractal.js, like this:
```
2. fractal.web.set('static.mount', '');
3. fractal.web.set('static.path', __dirname + '/assets/');
```

And run `fractal build`, it puts everything from the assets directory (in src directory) into the build directory.

If I run `fractal start` and go to localhost:3000 and click a component, I get the same output as before `No matching route found ...`.

So, in ALL CASES I cannot get `fractal start` to work as expected, and ONLY when having '/assets/' in BOTH places can I make `fractal build` work as expected.

---

Now, my root path is:
```
src
```

This is where my 'fractal.js' and 'gulpfile.js' etc. all lives.

And I define the path to my components like this:
```
...
3. fractal.components.set('path', __dirname + '/publish/node_modules');
...
```
*NOTE: Yes, that is the path, it's not a misspelling of 'public', it's the name of the app. And I use 'node_modules' to use the internal lookup machinery of nodes without having to mess around with adding additional paths to the lookup tree.*

```
src/
src/node_modules			< third party modules - ignored by git
src/publish/node_modules		< internal modules - included in git
```

Now, my preferred method of working and grouping files, assets etc. is to NOT have all my Markup, CSS and JavaScript in one single file – as seem very popular (for reasons unknown to me) in the front end community (is that even a thing?) at the momemt, as components tend to grow in more complex interfaces, you find yourself scrolling up and down A LOT –, but instead have my (s)css in one file, my markup (.hbs or .htm) in another and finally my JavaScript in a third. That way I can have them all open side by side, and scroll to the relevant place in each separate file at the same time, as that seem to be the most convenient when you either change or add functionality.

And when compiling, I can separate a fairly large and complex application into smaller separated and more focused applications (why load the whole application if you can't even log in? – serve the login app as a separate and very small app, and only the larger app if needed – yes, that is questioning the whole SPA thing).

This also means that I output my application parts into separate directories; leaving me with at structure like this:

```
assets/css		// global styles
assets/js		// global javascript
assets/libs		// global libraries
assets/login/css		// app specific styling
assets/login/js		// app specific javascript
assets/publish/css	// app specific styling
assets/publish/js	// app specific javascript
etc.
```

Each of these apps are based on an initial index.js file which feed data into a template and renders it. Each module can have from none to some different templates (a component can have various views).

This is to balance server requests vs. file caching – update the apps without having to serve new templates and vice versa.

In my case the output assets live on a level above the src folder like this:

```
publish/src
publish/static_dev
```

So I somehow have to point the output of Fractal into the 'static_dev' directory. Now, that was impossible for me to do, so I decided to move a copy of the static_dev into the src directory and renaming it 'assets' in my gulp build process. So now I DO have an assets folder INSIDE my src folder. And it also works just fine when I run `fractal build` – and view the index.html from within the output folder (providing the right `/assets/` in both places above).

The thing that throws me off is that it' not workin when running `fractal start`.

So inspecting the differences I can see that in the build version, in the iframe that loads the component, the assets are pointing to `../../loader/css/loader.css` whereas in the 'live' version it's pointing to `loader/css/loader.css` omitting the '../../' ... when I edit in the dev tools and add the '../../' it indeed works again. So somehow the live version omist these relative dot-slashes.

So, is this by design or a bug? And if by design, how do I get around it (what part am I missing/not getting)? I have tried adding all sort of variations of `../../` to the `static.mount` and `static.path` variables without any luck(!).

Now, stubborn as I am I'm not going to abandon this yet, as I really like the idea of being able to document your components from your actual source. I do find though, that having code doing fairly complicated stuff, often leads me to have to do a lot of cherry picking of code from my source and into the [component].config.js file. Just like I have to duplicate my Handlebars helpers from my code into fractal.js (using const hbs = require('@frctl/handlebars')({ ... });).

So, from an architectural point of view, there seems to be room for improvement (you probably already know that, so please don't see this as any kind of critique ; ) ). I'd like to have a 'system' that could automatically scan your code for helpers (handlebars or what have you, and adding them wherever appropriate for the system to work).

And therefore I thought: hey, why not fork it and see what I can do. Only to find the code completely stripped from comments (apart from a few couple scattered here and there) ... soooooh, kind a let that hang there ; )

And, please let me know if I can be of any help with this project, as I'd hate to see it disappear due to lack of resources.

**Comment by blaasvaer** [comment]
OK, found a workaround(!).

If I add `/assets/` to both the `static.mount` and `static.path`, AND add `../../assets/` to static files in the _preview.hbs BOTH `fractal build` and `fractal start` show the correct path.

Now, the problem is that I really don't get why they render the same now, and didn't with the others paths, and it's also a little confusing that the paths kind of lose relation with all this: to dot dot slash or not ...

**Comment by allmarkedup** [comment]
@blaasvaer ok, well at least you have it working now, that is a start :-)

Any paths you reference in the assets links in the template (i.e. ` ../../assets/css/foo.css`) are relative URL paths, not filesystem paths. All component previews are at URLs like `http://foo.com/components/preview/{component-name}` so the relative path to a css file that is served at `http://foo.com/assets/css/foo.css` from the preview template will be ` ../../assets/css/foo.css`. Not sure if that helps but just thought I'd point it out in case it was unclear!

Regarding your comments - they are all fair points and issues that we are definitely aware of. We tried to build Fractal to be as flexible as possible but there are so many different ways people want to organise component libraries and assets that it's very hard to find a solution that works for everything. Your example of using a components directory that contains a lot of independent NPM modules (with `node_modules` sub directories) is one situation that we did not plan for unfortunately and is not handled well by the current implementation. 

We are currently working pretty hard on Fractal v2 - a big focus on that will be a much higher quality of code, test coverage and it will be much more extensible than v1. As you have discovered, the code in the current version is really hard to work with and even for us to update unfortunately, which is why we have not been able to make many improvements to the current version. V2 is a big job however and is taking much longer than planned, but Fractal is certainly not disappearing :-)
